src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseStringEqualsOverStringCompare.Fixer.cs

```diff
@@ -0,0 +1,322 @@
+                            editor.ReplaceNode(operation.Syntax, replacementNode);
+                            return editor.GetChangedDocument();
+                        }, Resx.UseStringEqualsOverStringCompareCodeFixTitle);
+                    context.RegisterCodeFix(codeAction, context.Diagnostics);
+                    break;
+                }
+            }
+        }
+
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(UseStringEqualsOverStringCompare.RuleId);
+
+        public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        private static ImmutableArray<OperationSelector> GetSelectors(RequiredSymbols symbols)
+        {
+            return ImmutableArray.Create<OperationSelector>(
+                new StringStringSelector(symbols),
+                new StringStringBoolSelector(symbols),
+                new StringStringStringComparisonSelector(symbols));
+        }
+
+        /// <summary>
+        /// Gets a function that selects an operation that should have a diagnostic reported for it.
+        /// We have to use this round-about way of implementing the analyzer because <see cref="OperationSelector"/> has a method
+        /// that takes a <see cref="SyntaxGenerator"/>, and analyzers aren't allowed to reference that type in any way. Having the
+        /// analyzer use the <see cref="OperationSelector"/>s indirectly via this method avoids having duplicated logic in the analyzer and fixer.
+        /// </summary>
+        /// <param name="symbols"></param>
+        /// <returns></returns>
+        internal static Func<IOperation, bool> GetOperationSelector(RequiredSymbols symbols)
+        {
+            var selectors = GetSelectors(symbols);
+
+            return operation =>
+            {
+                foreach (var selector in selectors)
+                {
+                    if (selector.IsMatch(operation))
+                        return true;
+                }
+
+                return false;
+            };
+        }
+
+        /// <summary>
+        /// Selects an equals or not-equals operation where one argument is a literal zero, and the other argument
+        /// is an eligible <c>string.Compare</c> invocation.
+        /// </summary>
+        private abstract class OperationSelector
+        {
+            protected OperationSelector(RequiredSymbols symbols)
+            {
+                Symbols = symbols;
+            }
+
+            protected RequiredSymbols Symbols { get; }
+
+            /// <summary>
+            /// Indicates whether the specified <see cref="IOperation"/> matches the current <see cref="OperationSelector"/>
+            /// </summary>
+            /// <param name="compareResultToLiteralZero"></param>
+            /// <returns></returns>
+            public abstract bool IsMatch(IOperation compareResultToLiteralZero);
```

Comment:
Is it possible to change the parameter type to `IBinaryOperation` (and delegate the parameter type change to other methods as well?).

Reply:
I will do that when I restructure the analyzer.